### PR TITLE
feat: Jamba instruct tokenizer

### PR DESCRIPTION
**Pull request by Josephasafg**

### CODE REVIEW

**Comment by Josephasafg** [comment]
Notice that at the moment the vocabs path in huggingface is a placeholder until Jamba is officially released

File: .github/workflows/test.yaml [hunk 1]
```diff
@@ -44,7 +44,7 @@ jobs:
     runs-on: ubuntu-latest
     strategy:
       matrix:
-        python-version: ["3.7", "3.8", "3.9", "3.10", "3.11", "3.12"]
+        python-version: ["3.9", "3.10", "3.11", "3.12"]
```
Comment (etang-ai21) [inline .github/workflows/test.yaml:47]: Not even 3.8?

Comment (etang-ai21) [reply .github/workflows/test.yaml:47]: If we won't support these versions, I think it's worth a major version bump

Comment (etang-ai21) [reply .github/workflows/test.yaml:47]: Also does that mean that our SDK won't support python 3.8? Also might be a reason for a major version bump

File: ai21_tokenizer/tokenizer_factory.py [hunk 2]
```diff
@@ -25,23 +21,22 @@ class TokenizerFactory:
...
-        return JurassicTokenizer(model_path=model_path, config=config)
+    def get_tokenizer(cls, tokenizer_name: str = PreTrainedTokenizers.J2_TOKENIZER) -> BaseTokenizer:
+        if tokenizer_name == PreTrainedTokenizers.JAMBA_TOKENIZER:
+            return cls._create_jamaba_tokenizer("<huggingface-tokenizer-url-placeholder>")
 
-    @classmethod
-    def _tokenizer_dir(cls, tokenizer_name: str) -> Path:
-        return _LOCAL_RESOURCES_PATH / tokenizer_name
+        if tokenizer_name == PreTrainedTokenizers.J2_TOKENIZER:
+            return cls._create_jurassic_tokenizer(_LOCAL_RESOURCES_PATH / PreTrainedTokenizers.J2_TOKENIZER)
```
Comment (etang-ai21) [inline ai21_tokenizer/tokenizer_factory.py:30]: Worth adding a follow up task for removing this asset from the repo and downloading it same as jamba

File: ai21_tokenizer/tokenizer_factory.py [hunk 3]
```diff
@@ -1,22 +1,18 @@
...
 from ai21_tokenizer.base_tokenizer import BaseTokenizer
+from ai21_tokenizer.jamba_instruct_tokenizer import JambaInstructTokenizer
 from ai21_tokenizer.jurassic_tokenizer import JurassicTokenizer
-from ai21_tokenizer.utils import load_json
+from ai21_tokenizer.utils import PathLike
 
 _LOCAL_RESOURCES_PATH = Path(__file__).parent / "resources"
-_MODEL_EXTENSION = ".model"
-_MODEL_CONFIG_FILENAME = "config.json"
+_MODEL_CACHE_DIR = _LOCAL_RESOURCES_PATH / "cache"
```
Comment (etang-ai21) [inline ai21_tokenizer/tokenizer_factory.py:10]: Is that where HF are caching weights? in the source code?
Maybe add the option to override this with an env var

File: .github/workflows/test.yaml (hunk 1 above)
Comment (etang-ai21) [reply .github/workflows/test.yaml:47]: No tests for Jamba tokenizer?

File: ai21_tokenizer/tokenizer_factory.py (hunk 3 above)
Comment (Josephasafg) [reply ai21_tokenizer/tokenizer_factory.py:10]: I could use /tmp instead actually and use an env var. what do you think?

File: .github/workflows/test.yaml (hunk 1 above)
Comment (Josephasafg) [reply .github/workflows/test.yaml:47]: Regarding the version - there was an issue with numpy and the vast support of our Python versions. It was either 3.8 or 3.12

Comment (Josephasafg) [reply .github/workflows/test.yaml:47]: Regarding the tests - no. I'm waiting for the vocabs to be OS so I'll be able to test it.

**Review by etang-ai21 — APPROVED** [review]

**Review by miri-bar — APPROVED** [review]

File: ai21_tokenizer/jamba_instruct_tokenizer.py [hunk 4]
```diff
@@ -0,0 +1,76 @@
+from __future__ import annotations
+
+import os
+import tempfile
+from pathlib import Path
+from typing import Union, List, cast, Optional
+
+from tokenizers import Tokenizer
+from tokenizers import Tokenizer as TokenizerType
```
Comment (etang-ai21) [inline ai21_tokenizer/jamba_instruct_tokenizer.py:9]: why do you import this twice?

File: ai21_tokenizer/jamba_instruct_tokenizer.py [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+            model_path: str
+                The identifier of a Model on the Hugging Face Hub, that contains a tokenizer.json file
+            cache_dir: Optional[PathLike]
+                The directory to cache the tokenizer.json file.
+                 If not provided, the default cache directory will be used
+        """
+        self._tokenizer = self._init_tokenizer(model_path=model_path, cache_dir=cache_dir or _DEFAULT_MODEL_CACHE_DIR)
+
+    def _init_tokenizer(self, model_path: PathLike, cache_dir: PathLike) -> TokenizerType:
+        if Path(cache_dir).exists() and _TOKENIZER_FILE in os.listdir(cache_dir):
```
Comment (etang-ai21) [inline ai21_tokenizer/jamba_instruct_tokenizer.py:37]: extract to function (is_cached())

File: ai21_tokenizer/jamba_instruct_tokenizer.py [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+_DEFAULT_MODEL_CACHE_DIR = Path(tempfile.gettempdir()) / "tokenizer_cache"
+
+
+class JambaInstructTokenizer(BaseTokenizer):
+    _tokenizer: TokenizerType
+
+    def __init__(
+        self,
+        model_path: str,
+        cache_dir: Optional[PathLike] = _DEFAULT_MODEL_CACHE_DIR,
```
Comment (etang-ai21) [inline ai21_tokenizer/jamba_instruct_tokenizer.py:24]: Probably best this cache dir would include the jamba as a namespace (this way you can have different tokenizer caches)

Comment (etang-ai21) [reply ai21_tokenizer/jamba_instruct_tokenizer.py:24]: Here do:
`cache_dir = None`
in the first line of the init do `cache_dir = cacher_dir or _DEFAULT_MODEL_CACHE_DIR`

You got me wondering if you have a bug

File: ai21_tokenizer/jamba_instruct_tokenizer.py (hunk 4 above)
Comment (Josephasafg) [reply ai21_tokenizer/jamba_instruct_tokenizer.py:9]: Mistake. Removed

File: ai21_tokenizer/jamba_instruct_tokenizer.py (hunk 5 above)
Comment (Josephasafg) [reply ai21_tokenizer/jamba_instruct_tokenizer.py:37]: Done

File: ai21_tokenizer/jamba_instruct_tokenizer.py (hunk 6 above)
Comment (Josephasafg) [reply ai21_tokenizer/jamba_instruct_tokenizer.py]: Done. Thanks

**Review by etang-ai21 — APPROVED** [review]

**Pull request merged**